### PR TITLE
Add DopplerVelocityLogSystem plugin

### PR DESCRIPTION
**Pull request by hidmic**
# 🎉 New feature

## Summary

This patch adds a system plugin to pick DVL sensors. Needs https://github.com/gazebosim/gz-sensors/pull/290.

## Test it

TBD

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by hidmic** [comment]
First iteration. Once we've agreed on https://github.com/gazebosim/gz-sensors/pull/290, I'll go ahead and add integration tests.

**Comment by arjo129** [comment]
Check whitespaces.

**Comment by hidmic** [comment]
@iche033 @caguero I'm getting:

```
[Err] [SceneManager.cc:217] Visual: [horizontal_plane] already exists
237: [Err] [SceneManager.cc:217] Visual: [tethys] already exists
237: [Err] [BaseStorage.hh:927] Another item already exists with name: sun
```
in tests. Isn't the rendering engine cleaned up on test fixture destruction?

**Comment by iche033** [comment]
@osrf-jenkins run tests please

**Comment by iche033** [comment]
Made a few changes in [bdfe547](https://github.com/gazebosim/gz-sim/pull/1804/commits/bdfe5476295d9fb6975973c86e264bcd59f91f22) which should fix compilation.

There are a couple other issues left:
* `INTEGRATION_dvl_system` crashes. 
    *  if I run the tests individually, they do not crash but `DVLTest.BottomTracking` fails
* gz-sim core [adds a new dependency](https://github.com/gazebosim/gz-sim/blob/bdfe5476295d9fb6975973c86e264bcd59f91f22/src/CMakeLists.txt#L160) on gz-sensors
    *  I'm not sure if we want to do that. To workaround the issue, we would need to have a generic `EnvironmentalData` struct in gz sim and provide a function to convert between gz-sim and gz-sensors `EnvironmentalData` structs

**Comment by iche033** [comment]
> if I run the tests individually, they do not crash but `DVLTest.BottomTracking` and `DVLTest.WaterMassTracking` fail

* Worked around crash by splitting up tests into individual files. Currently rendering does not cleanup properly in between tests.
* Also fixed `DVLTest.BottomTracking` test
* `DVLTest.WaterMassTracking` is still failing. I'll probably need help from someone with knowledge on DVL for this. Test failure:

```
dvl_system_water_mass_tracking.cc:133: Failure
The difference between expectedLinearVelocityEstimate.X() and linearVelocityEstimate.X() is 0.50119416839480213, which exceeds kVelocityTolerance, where
expectedLinearVelocityEstimate.X() evaluates to -0.49920351475518465,
linearVelocityEstimate.X() evaluates to -1.0003976831499868, and
kVelocityTolerance evaluates to 0.10000000000000001.

/home/ian/code/gz_h_ws/src/gz-sim/test/integration/dvl_system_water_mass_tracking.cc:135: Failure
The difference between expectedLinearVelocityEstimate.Y() and linearVelocityEstimate.Y() is 1.4996016879686576, which exceeds kVelocityTolerance, where
expectedLinearVelocityEstimate.Y() evaluates to -1.0003978462872014,
linearVelocityEstimate.Y() evaluates to 0.49920384168145615, and
kVelocityTolerance evaluates to 0.10000000000000001.
```

**Comment by iche033** [comment]
> gz-sim core [adds a new dependency](https://github.com/gazebosim/gz-sim/blob/bdfe5476295d9fb6975973c86e264bcd59f91f22/src/CMakeLists.txt#L160) on gz-sensors

removed dependency on gz-sensors in core [edf0959](https://github.com/gazebosim/gz-sim/pull/1804/commits/edf0959ad76051ffe0695392b71f2bab330eee2c). One todo suggestion is to create the generic EnvironmentalData structure in sdformat and make gz-sensors and gz-sim use that data structure.

**Comment by iche033** [comment]
Made more fixes in [fbd8783](https://github.com/gazebosim/gz-sim/pull/1804/commits/fbd8783507bb96d6c7741bf21bd0eb7e14ade2d3)
* Fixed applying linear/angular velocity commands in body fixed frame (instead of world frame which has no effect) in Testfixture's ModelManipulator.
* Fixed DVL tests by computing the expected linear velocity in the correct reference frame
* **Disabled applying angular velocities to the AUV in tests**
    * If angular velocities were applied, the linear velocity estimates produced by the DVL sensor were found to be different from the expected values. I don't know if it's an issue with the test or an actual bug in the DVL sensor implementationin gz-sensors. The tests in gz-sensors also do not test this scenario. I added a todo note for this.

The tests are passing now with the caveat mentioned in the last bullet point.

**Comment by iche033** [comment]
ok I'm at the point that I think I could use a hand from people who are familiar with the DVL sensor implementation.

If I enable setting angular velocity on the AUV in the water mass tracking test by uncommenting [this line](https://github.com/gazebosim/gz-sim/blob/c0ef18f9e64ddbcebad017477d8e385014d1f7ed/test/integration/dvl_system_water_mass_tracking.cc#L124), the test fails:

```sh
/home/ian/code/gz_h_ws/src/gz-sim/test/integration/dvl_system_water_mass_tracking.cc:182: Failure
The difference between expectedLinearVelocityEstimate.X() and linearVelocityEstimate.X() is 0.31949900784986629, which exceeds kVelocityTolerance, where
expectedLinearVelocityEstimate.X() evaluates to -1.7163285450741428,
linearVelocityEstimate.X() evaluates to -1.3968295372242765, and
kVelocityTolerance evaluates to 0.10000000000000001.
/home/ian/code/gz_h_ws/src/gz-sim/test/integration/dvl_system_water_mass_tracking.cc:184: Failure
The difference between expectedLinearVelocityEstimate.Y() and linearVelocityEstimate.Y() is 1.023310004056881, which exceeds kVelocityTolerance, where
expectedLinearVelocityEstimate.Y() evaluates to -0.78063882907783577,
linearVelocityEstimate.Y() evaluates to -1.8039488331347169, and
kVelocityTolerance evaluates to 0.10000000000000001.
```

Maybe @arjo129 or @hidmic can shed some light on where or what the problem could be?

File: test/integration/dvl_system_water_mass_tracking.cc [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    math::Vector3d angularVelocityBodyFrame =
+        poses.back().Rot().RotateVectorReverse(
+            angularVelocities.back());
+    math::Vector3d angularVelocityRefFrame =
+        bodyToRef.RotateVectorReverse(angularVelocityBodyFrame);
+    math::Vector3d tangentialVelocityRefFrame =
+        angularVelocityRefFrame.Cross(sensorPositionInSFMFrame);
+    // calculate the final linear velocity estimate in reference frame
+    math::Vector3d expectedLinearVelocityEstimate =
+        linearVelocityRefFrame + tangentialVelocityRefFrame;
```
Comment (hidmic) [inline test/integration/dvl_system_water_mass_tracking.cc:178]: @iche033 this seems about right (thanks for the fix BTW, frames for angular velocity contributions were off). If the velocity commands are correct (see the request for clarification above), maybe sensor linear velocities w.r.t. the world frame as observed in the world frame back in the DVL are incorrect? [This code](https://github.com/gazebosim/gz-sensors/blob/0440bec8668da43f6fd84d8c9e5a14b3d679c94c/src/DopplerVelocityLog.cc#L1746-L1751) would be wrong if the reported velocity were not accounting for the transform between body and sensor frame.

**Review by hidmic — COMMENTED** [review]
@iche033 thank you for picking this up

File: test/helpers/ModelManipulator.hh [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        _ecm.CreateComponent(
+          model.Entity(), AngularVelocityCmd(
+            this->angularVelocityRequest.value()));
+      }
+      else
+      {
+        velocity->Data() = this->angularVelocityRequest.value();
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (iche033) [reply test/helpers/ModelManipulator.hh:89]: if reset, the velocities are only applied in that one iteration and moves the AUV only a little bit. I noticed that in the test comment, it is intended to make the AUV go in a circle so I removed the reset logic here in order to continuously apply the same velocities to the AUV.

File: test/helpers/ModelManipulator.hh [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  /// Update simulation state as necessary.
+  /// \note To be called on world pre-update.
+  /// \param[in] _ecm Entity component manager to update.
+  public: void Update(sim::EntityComponentManager &_ecm)
+  {
+    if (this->linearVelocityRequest.has_value())
+    {
+      sim::World world(sim::worldEntity(_ecm));
+      sim::Model model(world.ModelByName(_ecm, this->modelName));
+      using LinearVelocityCmd = sim::components::LinearVelocityCmd;
```
[the remark this answers is not in the document]
Comment (iche033) [reply test/helpers/ModelManipulator.hh:58]: I looked through the physics system and noticed that `WorldLinearVelocityCmd` is not being used at all. The `LinearVelocityCmd` applies the velocity in model frame, which I think it's what's intended after looking at the comment in the test. In order to move the AUV in circle, it needs to continuously apply x linear vel and z angular vel in the model frame

File: include/gz/sim/components/Environment.hh [hunk 4]
```diff
@@ -61,7 +61,15 @@ namespace components
...
-      bool _ignoreTimeStep = false);
+      bool _ignoreTimeStep = false)
+    {
+      auto data = std::make_shared<EnvironmentalData>();
+      data->frame = std::move(_frame);
+      data->reference = _reference;
+      data->units = _units;
+      data->staticTime = _ignoreTimeStep;
+      return data;
+    }
```
[the remark this answers is not in the document]
Comment (iche033) [reply include/gz/sim/components/Environment.hh:72]: this introduces gz-sensors dependency in the core component of gz-sim. So far, only the gz systems (plugins) have dependency on gz-sensors, which are are loaded at run time

The general approach we've been taking is to add the common data structure in sdformat in use them in other gz libraries. It's possible to add this extra dependency but I think it needs some discussion. I made the change to try and get the PR in first.

File: test/helpers/ModelManipulator.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply test/helpers/ModelManipulator.hh:89]: Oh, I see. So the command component themselves are reset on every iteration? That's good to know.

**Comment by iche033** [comment]
ok tests are now passing after changes in [a91ff9e](https://github.com/gazebosim/gz-sim/pull/1804/commits/a91ff9ecfc91a14f79b54d008fabe9e9ffaaa7f0)!

The maths are correct on the gz-sensors side. The problems were:
* Test computes the expected velocities using values from the wrong timestep. I updated the test to retrieve all model properties at the time that corresponds to the timestamp on the dvl message
* Ocassionally due to race condition between rendering and physics, there DVL system was passing a world state with incorrect timestamp to gz-sensors. Fixed by making sure SetWorldState is called right before render update.

This PR is ready.

**Comment by iche033** [comment]
CI is now working after mernging with `main`. Builds look good, merging!

**Review by iche033 — APPROVED** [review]

**Pull request merged**